### PR TITLE
feat(gatsby): add support for image-cdn

### PR DESCRIPTION
**Pull request by wardpeet**
## Description

Improve remote file handling by adding a new interface "RemoteFile" to GraphQL. This new interface allows source plugins to lazily download images and has native compatibility with gatsby-plugin-image.

To enable support for RemoteInterface, you'll need a couple of fields to be present on your node:
```
  url: `https://remote-url-to-file`,
  placeholderUrl: `https://remote-url-to-file?w=%width%&h=%height%&q=%quality%`,
  mimeType: `image/jpg`,
  filename: `name.jpg`,
  filesize: 10000,
  width: 800,
  height: 600,
```

As noted about only url, filename and mimeType are mandatory for regular files. Images also need width & height specified. The placeholderUrl field is a bit special as we added the option to give use a url with macros, so we can generate a smaller image to generate a placeholder to speed up query running.

With these nodes the RemoteFile Interface adds custom resolvers to your GraphQL type, without you needing to know how they work.
It adds publicUrl to generate a public facing url, resize to resize an image to a specific size and gatsbyImageData to give you a compatible data blob to use with gatsby-plugin-image.

---

The implementation is a bit hard to follow as everything is added inside `gatsby-plugin-utils` to support older gatsby v4 versions as well. You can upgrade gatsby-source-contentful without upgrading gatsby itself.

A source plugin can implement Gatsby's image CDN by wrapping the GraphQL type with the `addRemoteFilePolyfillInterface` in `createSchemaCustimization like this
```
import { addRemoteFilePolyfillInterface } from "gatsby-plugin-utils/polyfill-remote-file"

addRemoteFilePolyfillInterface(
  schema.buildObjectType({
    name: `PrefixAsset`,
    fields: {
      // your fields
    },
    interfaces: [`Node`, 'RemoteFile'],
  }),
  {
    schema, // schema you get as an argument from createSchemaCustomization
    store, // the redux store of gatsby also part of createSchemaCustomization
  }
)
```

Last but not least you'll need to add some extra routes to the dev server as well.
```
import { addRemoteFilePolyfillInterface } from "gatsby-plugin-utils/polyfill-remote-file"

addRemoteFilePolyfillInterface(
  schema.buildObjectType({
    name: `PrefixAsset`,
    fields: {
      // your fields
    },
    interfaces: [`Node`, 'RemoteFile'],
  }),
  {
    schema, // schema you get as an argument from createSchemaCustomization
    store, // the redux store of gatsby also part of createSchemaCustomization
  }
)
```

### CODE REVIEW

**Comment by TylerBarnes** [comment]
@wardpeet nice work! 🙌 The polyfill means it will work for users upgrading source plugins but staying on an old gatsby version. What about the other way around? Users upgrading gatsby but not source plugins. Will that break their existing GraphQL queries? I think we need to release a breaking change release for source plugins, but for gatsby we should probably put the new resolver changes under a flag or some other opt-in mechanism (perhaps a new gql field to make it opt-in?)

File: e2e-tests/development-runtime/package.json [hunk 1]
```diff
@@ -5,22 +5,22 @@
   "author": "Dustin Schau <dustin@gatsbyjs.com>",
   "dependencies": {
     "babel-plugin-search-and-replace": "^1.1.0",
-    "gatsby": "^4.6.0-next.4",
```
Comment (wardpeet) [inline e2e-tests/development-runtime/package.json:8]: Revert file

File: e2e-tests/production-runtime/package.json [hunk 2]
```diff
@@ -6,18 +6,18 @@
   "dependencies": {
     "babel-plugin-search-and-replace": "^1.1.0",
     "cypress": "^6.5.0",
-    "gatsby": "^4.1.6",
```
Comment (wardpeet) [inline e2e-tests/production-runtime/package.json:9]: Revert file

File: packages/gatsby-plugin-utils/src/polyfill-remote-file/jobs/dispatchers.ts [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+  ).split(`/`)
+
+  publicUrl.unshift(`public`)
+
+  const { actions } = importFrom(
+    global.__GATSBY?.root ?? process.cwd(),
+    `gatsby/dist/redux/actions`
+  ) as { actions: { createJobV2: (...args: any) => void } }
+
+  console.log({
```
Comment (wardpeet) [inline packages/gatsby-plugin-utils/src/polyfill-remote-file/jobs/dispatchers.ts:43]: Remove log

File: packages/gatsby-plugin-utils/src/polyfill-remote-file/placeholder-handler.ts [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+  originalHeight,
+}: {
+  url: string
+  width: number
+  originalWidth: number
+  originalHeight: number
+}): string {
+  const aspectRatio = originalWidth / originalHeight
+
+  return url
```
Comment (wardpeet) [inline packages/gatsby-plugin-utils/src/polyfill-remote-file/placeholder-handler.ts:277]: Add quality replacement too

File: e2e-tests/development-runtime/gatsby-node.js [hunk 5]
```diff
@@ -1,6 +1,75 @@
...
+      }),
+      {
+        store,
+        schema,
+      }
+    )
+  )
+}
+
+/** @type {imporg('gatsby').sourceNodes} */
```
Comment (LekoArts) [inline e2e-tests/development-runtime/gatsby-node.js:25]:
```suggestion
/** @type {import('gatsby').sourceNodes} */
```

File: e2e-tests/development-runtime/gatsby-node.js [hunk 6]
```diff
@@ -1,6 +1,75 @@
...
+exports.sourceNodes = ({ actions, createNodeId, createContentDigest }) => {
+  const items = [
+    {
+      name: "photoA.jpg",
+      url: "https://images.unsplash.com/photo-1517849845537-4d257902454a?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=2000&q=80",
```
Comment (LekoArts) [inline e2e-tests/development-runtime/gatsby-node.js:30]: Are those "stable" URLs that won't break our tests in the future?

File: docs/docs/how-to/testing/unit-testing.md [hunk 7]
```diff
@@ -44,6 +44,7 @@ module.exports = {
     ".+\\.(jpg|jpeg|png|gif|eot|otf|webp|svg|ttf|woff|woff2|mp4|webm|wav|mp3|m4a|aac|oga)$": `<rootDir>/__mocks__/file-mock.js`,
     "^gatsby-page-utils/(.*)$": `gatsby-page-utils/dist/$1`, // Workaround for https://github.com/facebook/jest/issues/9771
     "^gatsby-core-utils/(.*)$": `gatsby-core-utils/dist/$1`, // Workaround for https://github.com/facebook/jest/issues/9771
+    "^gatsby-plugin-utils/(.*)$": `gatsby-plugin-utils/dist/$1`, // Workaround for https://github.com/facebook/jest/issues/9771
```
Comment (LekoArts) [inline docs/docs/how-to/testing/unit-testing.md:47]: This isn't how we changed it in other places, so needs an update

File: packages/gatsby-plugin-utils/README.md [hunk 8]
```diff
@@ -76,7 +76,46 @@ Here's a list of features:
...
+if (!hasFeature(`image-cdn`)) {
+  // You can polyfill image-cdn here so older versions have support as well
+}
+```
+
+### Add ImageCDN support
+
+Our new ImageCDN allows source plugins to lazily download and process images. if you're a plugin author please use this polyfill to add support for all Gatsby V4 versions.
+
+For more information check (TODO)[]
```
Comment (LekoArts) [inline packages/gatsby-plugin-utils/README.md:88]: TODO here

File: packages/gatsby-plugin-utils/README.md [hunk 9]
```diff
@@ -76,7 +76,46 @@ Here's a list of features:
...
+
+#### Example
+
+```js
+const {
+  addRemoteFilePolyfillInterface,
+  polyfillImageServiceDevRoutes,
+} = require(`gatsby-plugin-utils/pollyfill-remote-file`)
+
+/** @type {import('gatsby').createSchemaCustomization} */
```
Comment (LekoArts) [inline packages/gatsby-plugin-utils/README.md:98]: We normally don't use type hints in our docs, I'd skip them

File: packages/gatsby-plugin-utils/package.json [hunk 10]
```diff
@@ -3,6 +3,31 @@
   "version": "3.3.0-next.0",
   "description": "Gatsby utils that help creating plugins",
   "main": "dist/index.js",
+  "exports": {
+    ".": "./dist/index.js",
+    "./*.js": "./dist/*.js",
+    "./dist/*": "./dist/*.js",
```
Comment (LekoArts) [inline packages/gatsby-plugin-utils/package.json:9]: Why do we need both mappings?

**Review by LekoArts — DISMISSED** [review]
Only managed to look at the first couple of files, need to look at it again tomorrow

File: packages/gatsby-plugin-utils/README.md (hunk 9 above)
Comment (wardpeet) [reply packages/gatsby-plugin-utils/README.md:98]: But they are soo good :joy: I'll remove it

File: e2e-tests/production-runtime/cypress/integration/remote-file.js [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+          height
+          width
+          src
+        }
+        fixed: gatsbyImageData(
+          layout: FIXED
+          width: 100
+          placeholder: DOMINANT_COLOR
+        )
+        constrained: gatsbyImageData(
```
[the remark this answers is not in the document]
Comment (wardpeet) [reply e2e-tests/production-runtime/cypress/integration/remote-file.js]: yes

File: e2e-tests/development-runtime/gatsby-node.js (hunk 6 above)
Comment (wardpeet) [reply e2e-tests/development-runtime/gatsby-node.js:30]: They should be stable yes

File: packages/gatsby-plugin-utils/package.json (hunk 10 above)
Comment (wardpeet) [reply packages/gatsby-plugin-utils/package.json:9]: To support both `gatsby-plugin-utils/dist/types` & `gatsby-plugin-utils/polyfill-remote-file`. Else you'll get `gatsby-plugin-utils/dist/dist/types`

File: packages/gatsby/src/schema/graphql-engine/lmdb-bundling-patch.ts [hunk 12]
```diff
@@ -16,20 +20,25 @@ import { createRequireFromPath } from "gatsby-core-utils"
 //  - https://github.com/DoctorEvidence/lmdb-js/blob/544b3fda402f24a70a0e946921e4c9134c5adf85/open.js#L77
 // Reliance on `import.meta.url` + usage of `.replace` is what seems to cause problems currently.
 
-export default function (source: string): string {
-  let lmdbBinaryLocation
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export default function (this: any, source: string): string {
```
Comment (wardpeet) [inline packages/gatsby/src/schema/graphql-engine/lmdb-bundling-patch.ts:24]: These changes are necessary as lmdb can also be required from gatsby-core-utils which can lead to multiple versions. Wich leads to different paths and we want to patch all lmdb versions that need patching.

File: packages/gatsby/src/schema/graphql-engine/bundle-webpack.ts [hunk 13]
```diff
@@ -71,7 +71,7 @@ export async function createGraphqlEngineBundle(
     module: {
       rules: [
         {
-          test: require.resolve(`lmdb`),
+          test: /node_modules[/\\]lmdb[/\\].*\.[cm]?js/,
```
Comment (wardpeet) [inline packages/gatsby/src/schema/graphql-engine/bundle-webpack.ts:74]: make the test reach other lmdb modules

File: packages/gatsby/src/schema/__tests__/node-model.js [hunk 14]
```diff
@@ -22,7 +22,7 @@ describe(`NodeModel`, () => {
     `SiteBuildMetadata`,
     `Author`,
     `Contributor`,
-    `RemoteFile`,
+    `ExternalFile`,
```
Comment (wardpeet) [inline packages/gatsby/src/schema/__tests__/node-model.js:25]: Changes necessary to pass tests, without it name conflicts happen

**Review by TylerBarnes — APPROVED** [review]

File: packages/gatsby-plugin-utils/src/polyfill-remote-file/index.ts [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+    )
+    store.dispatch(
+      // @ts-ignore - importFrom doesn't work with types
+      actions.createTypes(types, {
+        name: `gatsby`,
+        version: getGatsbyVersion(),
+        resolve: path.join(__dirname, `../`),
+      })
+    )
+  }
```
Comment (pieh) [inline packages/gatsby-plugin-utils/src/polyfill-remote-file/index.ts:123]: I think we should try to avoid importing from gatsby internals.

Can we instead of passing `store` and importing actions, pass `actions` from source plugins and "just" use 2nd param on action to tell it's `gatsby` that should create a type - like in https://github.com/gatsbyjs/gatsby/blob/2cf474b7c5d85ab4c62379c872f57f2cc07bd876/packages/gatsby-source-filesystem/src/create-remote-file-node.js#L76-L80 for file node creation

File: packages/gatsby-plugin-utils/src/polyfill-remote-file/placeholder-handler.ts [hunk 16]
```diff
@@ -0,0 +1,285 @@
...
+}
+
+const queue = Queue<
+  undefined,
+  {
+    url: string
+    contentDigest: string
+    width: number
+    height: number
+    hasCorrectFormat: boolean
```
Comment (wardpeet) [inline packages/gatsby-plugin-utils/src/polyfill-remote-file/placeholder-handler.ts:40]: Remove boolean because it's unused

File: packages/gatsby-plugin-utils/src/polyfill-remote-file/placeholder-handler.ts [hunk 17]
```diff
@@ -0,0 +1,285 @@
...
+  switch (type) {
+    case PlaceholderType.BLURRED: {
+      let buffer: Buffer
+
+      try {
+        const fileStream = createReadStream(filePath)
+        const pipeline = sharp()
+        fileStream.pipe(pipeline)
+        buffer = await pipeline
+          .resize(20, Math.ceil(20 / (width / height)))
```
Comment (wardpeet) [inline packages/gatsby-plugin-utils/src/polyfill-remote-file/placeholder-handler.ts:71]: Use width, height and original format

File: packages/gatsby-plugin-utils/src/polyfill-remote-file/index.ts (hunk 15 above)
Comment (wardpeet) [reply packages/gatsby-plugin-utils/src/polyfill-remote-file/index.ts:123]: I'll see if I can get it to  work

**Comment by nrandell** [comment]
Quick question on this. If you do a 'gatsby clean' on your development machine, does this delete the cache folder containing any processed images?

I love the feature though!

**Comment by TylerBarnes** [comment]
@nrandell it does 👍 On Gatsby Cloud images will be cached even when the site cache is invalidated but locally the images will be deleted. The nice thing is they're only fetched and processed once you visit localhost in `gatsby develop`, so it's not such a big deal if they're deleted
